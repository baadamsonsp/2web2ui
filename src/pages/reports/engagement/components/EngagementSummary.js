import React from 'react';
import { connect } from 'react-redux';
import { Grid, Panel } from '@sparkpost/matchbox';

import { PanelLoading } from 'src/components';
import { Percent } from 'src/components/formatters';
import { formatDateTime, relativeDateOptionsIndexed } from 'src/helpers/date';
import { formatFullNumber } from 'src/helpers/units';
import { safeRate } from 'src/helpers/math';
import styles from './EngagementSummary.module.scss';

export function EngagementSummary({ accepted = 0, clicks = 0, filters, loading, opens = 0, sent = 0 }) {
  if (loading) {
    return <PanelLoading minHeight="115px" />;
  }

  // sent should always be the largest number, so if it is zero all others should be zero
  if (sent === 0) {
    return null;
  }

  const timeRange = filters.relativeRange === 'custom'
    ? <span>from <strong>{formatDateTime(filters.from)}</strong> to <strong>{formatDateTime(filters.to)}</strong></span>
    : <span>in the <strong>{relativeDateOptionsIndexed[filters.relativeRange].toLowerCase()}</strong></span>;

  return (
    <Panel className={styles.EngagementSummary}>
      <Grid>
        <Grid.Column xs={12} md={3} xl={2}>
          <h1>< Percent value={safeRate(opens, accepted)} /></h1>
          <h6>Unique Open Rate</h6>
        </Grid.Column>
        <Grid.Column xs={12} md={3} xl={2}>
          <h1>< Percent value={safeRate(clicks, accepted)} /></h1>
          <h6>Unique Click Rate</h6>
        </Grid.Column>
        <Grid.Column xs={12} md={6} xl={8}>
          <p>
            Of <strong>{formatFullNumber(sent)}</strong> sent
            recipients, <strong>{formatFullNumber(accepted)}</strong> messages were
            accepted, <strong>{formatFullNumber(opens)}</strong> were opened
            and <strong>{formatFullNumber(clicks)}</strong> were clicked {timeRange}.
          </p>
        </Grid.Column>
      </Grid>
    </Panel>
  );
}

const mapStateToProps = ({ reportOptions: filters }) => ({ filters });
export default connect(mapStateToProps)(EngagementSummary);
