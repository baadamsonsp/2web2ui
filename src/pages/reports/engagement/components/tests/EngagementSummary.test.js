import React from 'react';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';
import moment from 'moment';

import { EngagementSummary } from '../EngagementSummary';

cases('EngagementSummary', (props) => {
  const wrapper = shallow(<EngagementSummary {...props} />);
  expect(wrapper).toMatchSnapshot();
}, {
  'renders loading panel': { loading: true },
  'returns null': { loading: false },
  'renders metric summary': {
    clicks: 123,
    filters: { relativeRange: 'hour' },
    loading: false,
    opens: 523,
    sent: 123123
  },
  'renders metric summary with custom time range': {
    clicks: 123,
    filters: {
      from: moment('2018-01-31T20:00:00'),
      relativeRange: 'custom',
      to: moment('2018-02-01T20:00:00')
    },
    loading: false,
    opens: 523,
    sent: 123123
  }
});
