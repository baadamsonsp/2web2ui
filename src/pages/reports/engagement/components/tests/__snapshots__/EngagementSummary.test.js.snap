// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngagementSummary renders loading panel 1`] = `
<PanelLoading
  minHeight="115px"
/>
`;

exports[`EngagementSummary renders metric summary 1`] = `
<Panel>
  <Grid>
    <Grid.Column
      md={3}
      xl={2}
      xs={12}
    >
      <h1>
        <Percent
          value={0.42477847355896137}
        />
      </h1>
      <h6>
        Unique Open Rate
      </h6>
    </Grid.Column>
    <Grid.Column
      md={3}
      xl={2}
      xs={12}
    >
      <h1>
        <Percent
          value={0.0999000999000999}
        />
      </h1>
      <h6>
        Unique Click Rate
      </h6>
    </Grid.Column>
    <Grid.Column
      md={6}
      xl={8}
      xs={12}
    >
      <p>
        Of 
        <strong>
          123,123
        </strong>
         sent recipients, 
        <strong>
          0
        </strong>
         messages were accepted, 
        <strong>
          523
        </strong>
         were opened and 
        <strong>
          123
        </strong>
         were clicked 
        <span>
          in the 
          <strong>
            last hour
          </strong>
        </span>
        .
      </p>
    </Grid.Column>
  </Grid>
</Panel>
`;

exports[`EngagementSummary renders metric summary with custom time range 1`] = `
<Panel>
  <Grid>
    <Grid.Column
      md={3}
      xl={2}
      xs={12}
    >
      <h1>
        <Percent
          value={0.42477847355896137}
        />
      </h1>
      <h6>
        Unique Open Rate
      </h6>
    </Grid.Column>
    <Grid.Column
      md={3}
      xl={2}
      xs={12}
    >
      <h1>
        <Percent
          value={0.0999000999000999}
        />
      </h1>
      <h6>
        Unique Click Rate
      </h6>
    </Grid.Column>
    <Grid.Column
      md={6}
      xl={8}
      xs={12}
    >
      <p>
        Of 
        <strong>
          123,123
        </strong>
         sent recipients, 
        <strong>
          0
        </strong>
         messages were accepted, 
        <strong>
          523
        </strong>
         were opened and 
        <strong>
          123
        </strong>
         were clicked 
        <span>
          from 
          <strong>
            Jan 31 2018, 8:00pm
          </strong>
           to 
          <strong>
            Feb 1 2018, 8:00pm
          </strong>
        </span>
        .
      </p>
    </Grid.Column>
  </Grid>
</Panel>
`;

exports[`EngagementSummary returns null 1`] = `""`;
